concent_api/core/tests/test_payments_report.py

```diff
@@ -0,0 +1,226 @@
+        for _ in range(self.number_of_clients_per_type):
+            client = Client(
+                public_key=base64.b64encode(ECCx(None).raw_pubkey)
+            )
+            client.full_clean()
+            client.save()
+            self.requestors.append(client)
+
+    def _prepare_subtasks(self):
+        for i in range(self.number_of_subtasks):
+            task_to_compute = self._get_deserialized_task_to_compute(
+                provider_public_key=self.providers[i // 3].public_key,
+                requestor_public_key=(
+                    self.requestors[(i + 1) // 3].public_key
+                    if i != (self.number_of_subtasks - 1) else self.requestors[0].public_key
+                ),
+            )
+            report_computed_task = self._get_deserialized_report_computed_task(task_to_compute=task_to_compute)
+
+            subtask = store_subtask(
+                task_id=report_computed_task.task_id,
+                subtask_id=report_computed_task.subtask_id,
+                provider_public_key=base64.b64decode(task_to_compute.provider_public_key),
+                requestor_public_key=base64.b64decode(task_to_compute.requestor_public_key),
+                state=Subtask.SubtaskState.ACCEPTED,
+                next_deadline=None,
+                task_to_compute=task_to_compute,
+                report_computed_task=report_computed_task,
+            )
+
+            self.subtasks.append(subtask)
+
+        assert Client.objects.count() == self.number_of_clients_per_type * 2
+
+    def _prepare_related_payments_mocks(self):
+        for subtask in get_subtasks_with_distinct_clients_pairs():  # pylint: disable=not-an-iterable
+            # The SCI methods are called in the same order in `core.payments.report.get_related_payments_for_pair`.
+            self.get_list_of_payments_mocks.append(
+                [
+                    self._create_batch_payment_object(
+                        self.default_payment_amount,
+                        self.closure_time,
+                    )
+                ]
+            )
+            self.get_list_of_payments_mocks.append(
+                [
+                    self._create_settlement_payment_object(
+                        self.default_payment_amount,
+                        self.closure_time,
+                    )
+                ]
+            )
+            self.get_list_of_payments_mocks.append(
+                [self._create_forced_subtask_payment_object(self.default_payment_amount, subtask.subtask_id)
+                 for subtask in Subtask.objects.filter(requestor=subtask.requestor, provider=subtask.provider)]
+            )
+            self.get_covered_additional_verification_costs_mocks.append(
+                [self._create_cover_additional_verification_costs_object(self.default_payment_amount, subtask.subtask_id)
+                 for subtask in Subtask.objects.filter(requestor=subtask.requestor, provider=subtask.provider)]
```

Comment:
this `for` loop is executed three times per iteration - list of related `subtask_ids` could be extracted to a variable and reused to produce:
`get_list_of_payments_mocks`,
`get_covered_additional_verification_costs_mocks`

Reply:
I've just extract that database call to variable 
```
subtask_list = Subtask.objects.filter(requestor=subtask.requestor, provider=subtask.provider)
``` 
and use it in those three for loops. I don't know if this is what You mean.

Reply:
actually, you could make list of `subtask_ids`:
```python
subtask_ids = list(subtask.id for subtask in Subtask.objects.filter(requestor=subtask.requestor, provider=subtask.provider))
```
so you don't have to store whole subtask in the memory, just their IDs (you don't need more for filtering)

Reply:
https://docs.djangoproject.com/en/2.1/ref/models/querysets/#values-list

Reply:
oh, it seems I need to catch up with django ;)